### PR TITLE
Implement nimbus generator

### PR DESCRIPTION
**Pull request by ductrandinh**
- Create `nimbus_generator` is a type conversion [Nimbus](https://github.com/nimbus-org/nimbus4flutter) client generator using [source_gen](https://github.com/dart-lang/source_gen) and inspired by [Chopper](https://github.com/lejard-h/chopper) and [Retrofit](https://github.com/square/retrofit).

- Update library

## Evidence

Generate code using the below command:
`dart run build_runner build --delete-conflicting-outputs`

https://github.com/nimbus-org/nimbus4flutter/assets/135354506/1ff20cbe-4718-4ed0-b69f-cbe3af568e14

### CODE REVIEW

**Comment by ductrandinh** [comment]
@m-takata  

Could you help me review and merge it?

And add me to Uploaders for nimbus4flutter on pub.dev

Screenshot 2023-10-02 at 13 38 11

File: lib/src/api/api_registory.dart [hunk 1]
```diff
@@ -133,7 +133,7 @@ class ApiRegistory{
   }
 
   /// Close repository.
-  static void close({bool force: false}){
+  static void close({bool force = false}){
```
Comment (m-takata) [inline lib/src/api/api_registory.dart:136]: In terms of the language specification, I think the meaning will be the same, but is the change made to unify the writing style?

File: nimbus_generator/example/lib/model/response/login_response_model.dart [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  factory LoginResponseRecord.fromJson(Map<String, dynamic> json) =>
+      _$LoginResponseRecordFromJson(json);
+
+  static RecordSchema? get schema => _$LoginResponseRecordSchema();
+
+  Map<String, dynamic> toJson() => _$LoginResponseRecordToJson(this);
+
+  final String? userId;
+  final String? sessionId;
+}
```
Comment (m-takata) [inline nimbus_generator/example/lib/model/response/login_response_model.dart]: Record is an extension of Map, and RecordList is an extension of List<Record>, but looking at the output of this generator, it seems that it is converted to a bean that contains a Record. Am I right?

File: nimbus_generator/example/lib/service/api_service.g.dart [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    );
+
+    ApiRegistory.registApi(
+      ApiHttp<Map<String, dynamic>, Map<String, dynamic>, Request, Response>(
+        name: 'login',
+        serverName: _kServerName,
+        method: HttpMethod.POST,
+        path: '/web/login.bf',
+        inputCreator: (_) => {},
+        outputCreator: (_) => {},
```
Comment (m-takata) [inline nimbus_generator/example/lib/service/api_service.g.dart:39]: In inputCreator and outputCreator, I assume that the API is used to create a DataSet association, but why is this empty and generated?

File: lib/src/api/api_registory.dart (hunk 1 above)
Comment (ductrandinh) [reply lib/src/api/api_registory.dart:136]: @m-takata  Currently, it still works, but in the future, it won't be supported. You can read it here

https://dart.dev/tools/diagnostic-messages?utm_source=dartdev&utm_medium=redir&utm_id=diagcode&utm_content=obsolete_colon_for_default_value#obsolete_colon_for_default_value

File: nimbus_generator/example/lib/service/api_service.g.dart (hunk 3 above)
Comment (ductrandinh) [reply nimbus_generator/example/lib/service/api_service.g.dart:39]: I still use a DataSet association

this concept is difficult to learn and developers are already familiar with how to use json_serializable so I wrote a similar library inspired json_serializable  by using source_gen.

In the login function of the api_service.g.dart file, you can see two functions toJson and fromJson, that will auto-convert to Dataset and return the expected result. 

toJson
Model -> Dataset -> Map 
fromJson
Map -> Dataset -> Model

File: nimbus_generator/example/lib/model/response/login_response_model.dart (hunk 2 above)
Comment (ductrandinh) [reply nimbus_generator/example/lib/model/response/login_response_model.dart]: Yes, it should be the schema, I will update them

Comment (ductrandinh) [reply nimbus_generator/example/lib/model/response/login_response_model.dart]: I renamed Record to Schema

File: lib/src/api/api_registory.dart (hunk 1 above)
Comment (m-takata) [reply lib/src/api/api_registory.dart:136]: Okay, I understand.

File: nimbus_generator/example/lib/model/response/login_response_model.dart (hunk 2 above)
Comment (m-takata) [reply nimbus_generator/example/lib/model/response/login_response_model.dart]: One of the goals of this library is to achieve dynamic DTOs by defining a schema in the first place, without having to develop concrete classes to define input and output DTOs.

So, defining a static DTO class and then generating a class that encompasses the dynamic DTOs of this library by adding decorations to it is in itself an act that hinders the purpose of the library.

Therefore, I would like to hear the intent of this.

File: nimbus_generator/example/lib/service/api_service.g.dart (hunk 3 above)
Comment (m-takata) [reply nimbus_generator/example/lib/service/api_service.g.dart:39]: The purpose is to increase code readability by declaring input DTO definitions and output DTO definitions in inputCreator and outputCreator in a declaration called API.
Therefore, an empty implementation of inputCreator and outputCreator is not the design intent of this library.

**Comment by m-takata** [comment]
I understand your intention to increase development productivity by using code generation.

I also understand the intention to map DataSet by decorating the bean, since the concept of DataSet is difficult to understand.

However, it seems to me that you are trying to create code generation without understanding the original purpose of this library.

I think you should first write some code that should be implemented when the library is used purely, and then take the step of making the code generation more efficient where it can be made more efficient.

**Comment by ductrandinh** [comment]
> I understand your intention to increase development productivity by using code generation.
> 
> I also understand the intention to map DataSet by decorating the bean, since the concept of DataSet is difficult to understand.
> 
> However, it seems to me that you are trying to create code generation without understanding the original purpose of this library.
> 
> I think you should first write some code that should be implemented when the library is used purely, and then take the step of making the code generation more efficient where it can be made more efficient.


We used the Nimbus4Flutter library for the Invest Point project (I worked with YukiKoyamatsu) and I created a demo app using API login, I feel I must write duplicate code, it is really not necessary.
With the base source I used in the project, I don't need dynamic DTOs. I trust with the current code most of the developers can easily understand because they have worked with the same concept before.

With this pull request, I write a new library and I don't change the logic of any code in your nimbus4flutter library. The developers can choose what they want your nimbus4flutter or my annotation generator code
 
I understand the library must adapt to multi-case but currently, we need to finish the base API task for the current project on the last day of the week, temporarily I don't have time to follow your opinion. I will update it later.

What do you think about it?

**Comment by m-takata** [comment]
It is unacceptable to merge PRs in the current halfway state.

It is too early to determine if this code generation is usable, so please commit it to your project's git repository first.

**Comment by ductrandinh** [comment]
OK, I will create a new project and push it

**Pull request closed without merging**